File: distbench-py/distbench/algorithms/echo.py
"""Echo algorithm implementation

A simple distributed algorithm where a start node sends messages to all
other nodes, which echo them back.
"""

import asyncio
import logging
from typing import Any

from distbench.algorithm import Algorithm
from distbench.community import PeerId
from distbench.decorators import config_field, distbench, handler, message

logger = logging.getLogger(__name__)


@message
class Message:
    """A message to be echoed by peers."""

    sender: str
    message: str


@distbench
class Echo(Algorithm):
    """Echo algorithm implementation.

    If start_node is True, this node initiates by sending messages to all peers.
    All nodes echo received messages back to the sender.
    """

    start_node: bool = config_field(default=False)

    def __init__(self, config: dict[str, object], peers: dict[PeerId, Any]) -> None:
        """Initialize Echo algorithm."""
        super().__init__()
        self.messages_received = 0

    async def on_start(self) -> None:
        """Called when algorithm starts.

        If this is the start node, send messages to all peers.
        """
        logger.info("Echo algorithm starting")

        if self.start_node:
            logger.info("This is the start node, sending messages to all peers")
            msg = Message(sender="Test", message="Hello, world!")

            for peer_id, peer in self.peers.items():
                try:
                    logger.info(f"Sending message to {peer_id}")
                    # Call the message handler on the peer
                    response = await peer.message(msg)
                    if response:
                        logger.info(f"Message echoed back: {response}")
                    else:
                        logger.error(f"Message not echoed by {peer_id}")

                except Exception as e:
                    logger.error(f"Error echoing message to {peer_id}: {e}")

        # Wait a bit for all messages to complete
        await asyncio.sleep(1.0)

        # Terminate the algorithm
        await self.terminate()

    async def on_exit(self) -> None:
        """Called when algorithm exits."""
        logger.info("Echo algorithm exiting")

    async def report(self) -> dict[str, str]:
        """Generate algorithm report.

        Returns:
            Dictionary with metrics
        """
        return {
            "messages_received": str(self.messages_received),
        }

    @handler
    async def message(self, src: PeerId, msg: Message) -> str | None:
        """Handle an incoming echo message.

        Args:
            src: Sender's peer ID
            msg: Message to echo

        Returns:
            String echoing the message content
        """
        logger.info(f"Received message from {src}: {msg.message}")
        self.messages_received += 1
        return msg.message

File: distbench-rs/src/algorithms/echo.rs
use std::{
    collections::HashMap,
    fmt::Display,
    sync::atomic::{AtomicU64, Ordering},
    time::Duration,
};

use async_trait::async_trait;
use common_macros::hash_map;
use distbench::{self, community::PeerId, Algorithm, SelfTerminating};
use log::{error, info};

#[distbench::message]
struct Message {
    sender: String,
    message: String,
}

#[distbench::state]
pub struct Echo {
    #[distbench::config(default = false)]
    start_node: bool,
    messages_received: AtomicU64,
}

#[async_trait]
impl Algorithm for Echo {
    async fn on_start(&self) {
        info!("Echo algorithm starting");
        if self.start_node {
            for (_, peer) in self.peers() {
                match peer
                    .message(&Message {
                        sender: "Test".to_string(),
                        message: "Hello, world!".to_string(),
                    })
                    .await
                {
                    Ok(Some(message)) => info!("Message echoed: {}", message),
                    Ok(None) => error!("Message not echoed"),
                    Err(e) => error!("Error echoing message: {}", e),
                }
            }
        }
        tokio::time::sleep(Duration::from_secs(1)).await;
        self.terminate().await;
    }

    async fn on_exit(&self) {
        info!("Echo algorithm exiting");
    }

    async fn report(&self) -> Option<HashMap<impl Display, impl Display>> {
        Some(hash_map! {
            "messages_received" => self.messages_received.load(Ordering::Relaxed).to_string(),
        })
    }
}

#[distbench::handlers]
impl Echo {
    async fn message(&self, src: PeerId, msg: &Message) -> Option<String> {
        info!("Received message from {}: {}", src.to_string(), msg.message);
        self.messages_received.fetch_add(1, Ordering::Relaxed);
        Some(msg.message.clone())
    }
}
